src/ripple/app/paths/impl/PaySteps.cpp

```diff
@@ -0,0 +1,533 @@
+        return make_XRPEndpointStep (ctx, e2->getAccountID());
+
+    if (e1->isAccount() && e2->isAccount())
+    {
+        return make_DirectStepI (ctx, e1->getAccountID (),
+            e2->getAccountID (), curIssue.currency);
+    }
+
+    if (e1->isOffer() && e2->isAccount())
+    {
+        // should already be taken care of
+        JLOG (j.warning)
+            << "Found offer/account payment step. Aborting payment strand.";
+        assert (0);
+        Throw<FlowException> (tefEXCEPTION, "Found offer/account payment step.");
+    }
+
+    assert ((e2->getNodeType () & STPathElement::typeCurrency) ||
+        (e2->getNodeType () & STPathElement::typeIssuer));
+    auto const outCurrency = e2->getNodeType () & STPathElement::typeCurrency
+        ? e2->getCurrency ()
+        : curIssue.currency;
+    auto const outIssuer = e2->getNodeType () & STPathElement::typeIssuer
+        ? e2->getIssuerID ()
+        : curIssue.account;
+
+    if (isXRP (curIssue.currency) && isXRP (outCurrency))
+    {
+        JLOG (j.warning) << "Found xrp/xrp offer payment step";
+        return {temBAD_PATH, std::unique_ptr<Step>{}};
+    }
+
+    assert (e2->isOffer ());
+
+    if (isXRP (outCurrency))
+        return make_BookStepIX (ctx, curIssue);
+
+    if (isXRP (curIssue.currency))
+        return make_BookStepXI (ctx, {outCurrency, outIssuer});
+
+    return make_BookStepII (ctx, curIssue, {outCurrency, outIssuer});
+}
+
+std::pair<TER, Strand>
+toStrand (
+    ReadView const& view,
+    AccountID const& src,
+    AccountID const& dst,
+    Issue const& deliver,
+    boost::optional<Issue> const& sendMaxIssue,
+    STPath const& path,
+    beast::Journal j)
+{
+    if (isXRP (src))
+    {
+        JLOG (j.debug) << "toStrand with xrpAccount as src";
+        return {temBAD_PATH, Strand{}};
+    }
+    if (isXRP (dst))
+    {
+        JLOG (j.debug) << "toStrand with xrpAccount as dst";
+        return {temBAD_PATH, Strand{}};
+    }
+    if (!isConsistent (deliver))
+    {
+        JLOG (j.debug) << "toStrand inconsistent deliver issue";
+        return {temBAD_PATH, Strand{}};
+    }
+    if (sendMaxIssue && !isConsistent (*sendMaxIssue))
+    {
+        JLOG (j.debug) << "toStrand inconsistent sendMax issue";
+        return {temBAD_PATH, Strand{}};
+    }
+
+    Issue curIssue = [&]
+    {
+        auto& currency =
+            sendMaxIssue ? sendMaxIssue->currency : deliver.currency;
+        if (isXRP (currency))
+            return xrpIssue ();
+        return Issue{currency, src};
+    }();
+
+    STPathElement const firstNode (
+        STPathElement::typeAll, src, curIssue.currency, curIssue.account);
+
+    boost::optional<STPathElement> sendMaxPE;
+    if (sendMaxIssue && sendMaxIssue->account != src)
+        sendMaxPE.emplace (sendMaxIssue->account, boost::none, boost::none);
+
+    STPathElement const lastNode (dst, boost::none, boost::none);
+
+    auto isAccount = [](STPathElement const* pe)
```

Comment:
I got rid of all of these except `hasCurrency`, which I only used as an argument to `find_if`. I replaced the other uses with a member function on `STPathElement`.